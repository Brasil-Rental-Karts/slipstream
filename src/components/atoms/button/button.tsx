import * as React from "react";
import { Slot } from "@radix-ui/react-slot";
import { cva, type VariantProps } from "class-variance-authority";
import { cn } from "../../../lib/cn";

const buttonVariants = cva(
  "inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-lg font-medium transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-60 select-none",
  {
    variants: {
      variant: {
        default: "bg-primary text-white shadow-soft hover:bg-primary/90 active:bg-primary/80",
        secondary: "bg-secondary text-white hover:bg-secondary/90 active:bg-secondary/80",
        outline: "border border-foreground/20 text-foreground bg-transparent hover:bg-foreground/5 active:bg-foreground/10",
        ghost: "bg-transparent text-foreground hover:bg-foreground/5 active:bg-foreground/10",
        destructive: "bg-destructive text-white hover:bg-destructive/90 active:bg-destructive/80",
      },
      size: {
        sm: "h-8 px-3 text-sm",
        md: "h-10 px-4 text-sm",
        lg: "h-12 px-6 text-base",
      },
      loading: {
        true: "cursor-progress",
      },
    },
    defaultVariants: {
      variant: "default",
      size: "md",
    },
  }
);

export interface ButtonProps
  extends React.ButtonHTMLAttributes<HTMLButtonElement>,
    VariantProps<typeof buttonVariants> {
  asChild?: boolean;
  isLoading?: boolean;
}

export const Button = React.forwardRef<HTMLButtonElement, ButtonProps>(
  ({ className, variant, size, asChild = false, isLoading = false, disabled, children, ...props }, ref) => {
    const Comp = asChild ? Slot : "button";
    const isDisabled = disabled || isLoading;
    return (
      <Comp
        className={cn(buttonVariants({ variant, size, loading: isLoading ? true : undefined }), className)}
        ref={ref}
        disabled={isDisabled}
        {...props}
      >
        {isLoading && (
          <svg
            className="animate-spin h-4 w-4 text-current"
            xmlns="http://www.w3.org/2000/svg"
            fill="none"
            viewBox="0 0 24 24"
            aria-hidden="true"
          >
            <circle className="opacity-25" cx="12" cy="12" r="10" stroke="currentColor" strokeWidth="4"></circle>
            <path className="opacity-75" fill="currentColor" d="M4 12a8 8 0 018-8v4a4 4 0 00-4 4H4z"></path>
          </svg>
        )}
        {children}
      </Comp>
    );
  }
);
Button.displayName = "Button";

export { buttonVariants };


