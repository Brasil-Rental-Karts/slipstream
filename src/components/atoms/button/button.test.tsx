import * as React from 'react';
import { describe, it, expect, vi } from 'vitest';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Button } from './button';

describe('Button', () => {
  it('renderiza com variant default', () => {
    render(<Button>Label</Button>);
    const btn = screen.getByRole('button', { name: /label/i });
    expect(btn).toBeInTheDocument();
  });

  it('renderiza com variant secondary', () => {
    render(<Button variant="secondary">Secondary</Button>);
    expect(screen.getByRole('button', { name: /secondary/i })).toBeInTheDocument();
  });

  it('renderiza com variant outline', () => {
    render(<Button variant="outline">Outline</Button>);
    expect(screen.getByRole('button', { name: /outline/i })).toBeInTheDocument();
  });

  it('renderiza com variant ghost', () => {
    render(<Button variant="ghost">Ghost</Button>);
    expect(screen.getByRole('button', { name: /ghost/i })).toBeInTheDocument();
  });

  it('renderiza com variant destructive', () => {
    render(<Button variant="destructive">Delete</Button>);
    expect(screen.getByRole('button', { name: /delete/i })).toBeInTheDocument();
  });

  it('renderiza tamanhos sm, md, lg', () => {
    render(
      <div>
        <Button size="sm">Small</Button>
        <Button size="md">Medium</Button>
        <Button size="lg">Large</Button>
      </div>
    );
    expect(screen.getByRole('button', { name: /small/i })).toBeInTheDocument();
    expect(screen.getByRole('button', { name: /medium/i })).toBeInTheDocument();
    expect(screen.getByRole('button', { name: /large/i })).toBeInTheDocument();
  });

  it('dispara onClick quando clicado', async () => {
    const user = userEvent.setup();
    const onClick = vi.fn();
    render(<Button onClick={onClick}>Click</Button>);
    await user.click(screen.getByRole('button', { name: /click/i }));
    expect(onClick).toHaveBeenCalledTimes(1);
  });

  it('exibe estado de loading com spinner e desabilita clique', async () => {
    const user = userEvent.setup();
    const onClick = vi.fn();
    render(<Button isLoading onClick={onClick}>Loading</Button>);
    const btn = screen.getByRole('button', { name: /loading/i });
    expect(btn).toHaveAttribute('disabled');
    // SVG spinner está presente
    expect(btn.querySelector('svg')).toBeTruthy();
    await user.click(btn);
    expect(onClick).not.toHaveBeenCalled();
  });

  it('aplica disabled corretamente', async () => {
    const user = userEvent.setup();
    const onClick = vi.fn();
    render(<Button disabled onClick={onClick}>Disabled</Button>);
    const btn = screen.getByRole('button', { name: /disabled/i });
    expect(btn).toBeDisabled();
    await user.click(btn);
    expect(onClick).not.toHaveBeenCalled();
  });

  it('é acessível via role button e aceita aria-*', () => {
    render(<Button aria-label="custom" aria-live="polite">A11y</Button>);
    const btn = screen.getByRole('button', { name: /custom/i });
    expect(btn).toHaveAttribute('aria-label', 'custom');
    expect(btn).toHaveAttribute('aria-live', 'polite');
  });

  it('snapshot básico', () => {
    const { container } = render(<Button variant="secondary" size="lg">Snapshot</Button>);
    expect(container.firstChild).toMatchSnapshot();
  });
});


