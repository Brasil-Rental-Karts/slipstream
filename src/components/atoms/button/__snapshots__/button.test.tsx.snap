// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button > snapshot b√°sico 1`] = `
<button
  class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-lg font-medium transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-60 select-none bg-secondary text-white hover:bg-secondary/90 active:bg-secondary/80 h-12 px-6 text-base"
>
  Snapshot
</button>
`;
